hs_restclient/__init__.py

```diff
@@ -1 +1,666 @@
+        :param keywords: list of strings representing keywords to associate with the resource
+        :param edit_users: list of HydroShare usernames who will be given edit permissions
+        :param view_users: list of HydroShare usernames who will be given view permissions
+        :param edit_groups: list of HydroShare group names that will be given edit permissions
+        :param view_groups: list of HydroShare group names that will be given view permissions
+
+        :return: string representing ID of newly created resource.
+
+        :raises: HydroShareArgumentException if any parameters are invalid.
+        :raises: HydroShareNotAuthorized if user is not authorized to perform action.
+        :raises: HydroShareHTTPException if an unexpected HTTP response code is encountered.
+
+        """
+        url = "{url_base}/resource/".format(url_base=self.url_base)
+
+        close_fd = False
+
+        if not resource_type in self.resource_types:
+            raise HydroShareArgumentException("Resource type {0} is not among known resources: {1}".format(resource_type,
+                                                                                                           ", ".join([r for r in self.resource_types])))
+        files = None
+        if resource_file:
+            (files, close_fd) = self._prepareFileForUpload(resource_file, resource_filename)
+
+        # Prepare request
+        params = {'resource_type': resource_type, 'title': title}
+        if abstract:
+            params['abstract'] = abstract
+        if keywords:
+            # Put keywords in a format that django-rest's serializer will understand
+            for (i, kw) in enumerate(keywords):
+                key = "keywords[{index}]".format(index=i)
+                params[key] = kw
+        if edit_users:
+            params['edit_users'] = edit_users
+        if view_users:
+            params['view_users'] = view_users
+        if edit_groups:
+            params['edit_groups'] = edit_groups
+        if view_groups:
+            params['view_groups'] = view_groups
+
+        r = self._request('POST', url, data=params, files=files)
+
+        if close_fd:
+            fd = files['file'][1]
+            fd.close()
+
+        if r.status_code != 201:
+            if r.status_code == 403:
+                raise HydroShareNotAuthorized(('POST', url))
+            else:
+                raise HydroShareHTTPException((url, 'POST', r.status_code, params))
+
+        response = r.json()
+        new_resource_id = response['resource_id']
+
+        if response['resource_type'] != resource_type:
+            msg = "New resource {resource_id} was created, but the new resource type is {new_type}, " + \
+                  "while the expected type is {exp_type}."
+            msg = msg.format(resource_id=new_resource_id, new_type=response['resource_type'], exp_type=resource_type)
+            raise HydroShareException(msg)
```

Comment:
An exception, even though a resource was created? How is a user supposed to handle that properly?